src/executioners/OptimizeSolve.C

```diff
@@ -263,16 +261,47 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
 }
 
 PetscErrorCode
-OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat hessian, Mat /*pc*/, void * ctx)
+OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat /*hessian*/, Mat /*pc*/, void * ctx)
 {
   auto * solver = static_cast<OptimizeSolve *>(ctx);
-
   libMesh::PetscVector<Number> param(x, solver->_my_comm);
   *solver->_parameters = param;
+  auto n = solver->_ndof;
+  auto comm = solver->_my_comm.get();
+  MatCreateShell(comm, n,n,n,n, ctx, &(solver->_hessian)); // need to fix two of the n's for parallel
+  MatShellSetOperation(solver->_hessian, MATOP_MULT, (void(*)(void)) OptimizeSolve::applyHessianWrapper);
+  return 0;
+}
 
-  libMesh::PetscMatrix<Number> mat(hessian, solver->_my_comm);
+PetscErrorCode
+OptimizeSolve::applyHessianWrapper(Mat H, Vec s, Vec Hs)
+{
+  void *ctx;
+  MatShellGetContext(H,&ctx);
+  auto * solver = static_cast<OptimizeSolve *>(ctx);
+  libMesh::PetscVector<Number> sbar(s, solver->_my_comm);
+  libMesh::PetscVector<Number> Hsbar(Hs, solver->_my_comm);
+  return solver->applyHessian(sbar,Hsbar);
+}
 
-  solver->hessianFunction(mat);
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> &s, libMesh::PetscVector<Number> &Hs)
+{
+  _form_function->updateParameters(s); 
```

Comment:
This is where the _parameters are gotten from the solver with the other wrappers using:
```_form_function->updateParameters(*_parameters.get());```